### ISSUE TITLE
PVC resource limits not validated

### ISSUE THREAD

**Issue by msau42** [kubernetes/kubernetes#106637]
### What happened?

PVC never supported setting capacity limits. However, it looks like in validation, we never enforced it: https://github.com/kubernetes/kubernetes/blob/a5622f3f6e1f5d5ffd066d799204ea0c6d1a49e2/pkg/apis/core/validation/validation.go#L2124

This means that it's possible PVC.spce.resource.limits could have garbage values in them. 

### What did you expect to happen?

Validation should have rejected PVC setting resource.limits. 

If we want to tighten validation, we would need to think about how to do it carefully for existing objects. Also, we should consider if we want to support limits.

### How can we reproduce it (as minimally and precisely as possible)?

Create a PVC and set resource.limits

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
```

</details>

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by NikhilSharmaWe** [comment]
@msau42 I would like to work on this issue. I am not able to find any code or info about resource.limits object. Could you please provide some resources, so that I can better my understanding about what we are trying to achieve here.

**Comment by jlsong01** [comment]
@NikhilSharmaWe  Hello NikhilSharmaWe, are you working on this issue? If not, could I take a try?

**Comment by NikhilSharmaWe** [comment]
@jlsong01 I am unable to figure out how to make object `PVC.spce.resource.limits` free of garbage values because I am unable to find any info about this object. @msau42 Could you guide about this object and some useful code pointers.

@jlsong01 if you have an idea of making a useful solution, you can very well proceed. Also if you have some idea about how to execute this task, could you please share with me. Maybe, we can collaborate.

**Comment by jlsong01** [comment]
Here is the fix proposal, please feel free to comment if anything needed. @msau42 
1. The tighten validation would be added, which means, if `spec.Resources.Limits[core.ResourceStorage]` is present, not matter what value in them, the validation will failed with error indication.
2. Regarding pvc supports limit or not , IMHO, No. Since the storage claimed by pvc can not be shared with another pvc, the pvc limit may not work as the request&limits of CPU or Memory

This would be a small size PR, @NikhilSharmaWe maybe you can help to review the above idea and the following PR later.

**Comment by mattcary** [comment]
I don't think we can just add validation, because there may be resources out there that have set the limit, and we don't want to break those clusters.

If we know that we want to start using limits, eg plumb them through to the csi driver, then we want to start validating the the limit is a valid storage quantity. If we don't have a plan for using limits, then it's not as clear to me what to do---perhaps start warning, in order to make it easier to add some features in the future.

The other dimension to think about is how to validate. Adding API validation is problematic, as mentioned, because it could break existing clusters. I think that means the option is to add a webhook that will warn on invalid PVC, and then we can ratchet that to enforcement over several k8s versions, and only then move this into api validation.

Or are there other options here?

**Comment by mattcary** [comment]
(fyi, it looks like key people are out during the holiday time, so responses might be delayed)

**Comment by jlsong01** [comment]
copy from PR: https://github.com/kubernetes/kubernetes/pull/107077, which was closed somehow

@mattcary @jan-safranek Thanks for your comments. Here is my work plan, maybe it's another option, could you please take a look?


The tightened validation: If spec.Resources.Limits[core.ResourceStorage] is present, no matter what value in them, the validation will fail

-  add this tightened validation on create

-  add this tightened validation on update where the old PersistentVolumeClaimSpec passed this validation

-  skip this tightened validation on update where the old PersistentVolumeClaimSpec fails this validation

Unit test will be handled acoordingly.

**Comment by mattcary** [comment]
Sorry for the delay!

Looks like this is exactly [Jordan's guidance](https://github.com/kubernetes/kubernetes/pull/103024#issuecomment-874846127), so I think this plan makes sense. Go for it!

Thanks for your work on this.

**Comment by jlsong01** [comment]
Thanks for the confirmation. The work is in progress...

**Comment by mattcary** [comment]
/remove lifecycle-stale
/remove-lifecycle stale

@jlsong01 is still working on this.

**Comment by jlsong01** [comment]
@mattcary Hey, any progress on the PR review? Please let me know if anything needed from my side. Thanks :-)

**Comment by msau42** [comment]
The tricky thing here is that there could be external controllers that are using/depending on these fields. So I think the best things we can do for now are:

- Allow existing objects that are invalid to remain invalid
- Validate that on create/update the `limit("storage"]` values are valid (>= request, well-formed)
- When set, add a warning that limits are ignored by Kubernetes
- When set, add a warning that any request other than `storage` are ignored by Kubernetes

**Comment by xing-yang** [comment]
Another question is should we support resource.limits?
